### ISSUE TITLE
Large memory requirements in ewald calculation

### ISSUE THREAD

**Issue by willwheelera** [WagnerGroup/pyqmc#159]
I'm running line_minimization on a large system (64 Si atoms) using the following setup

    #!/usr/bin/env python

    import numpy as np
    import pyqmc
    from pyqmc.dasktools import distvmc as vmc, line_minimization
    from dask.distributed import Client, LocalCluster
    import generate

    if __name__=="__main__":
        slater = generate.get_slater(S=2 * (np.ones((3,3)) - 2 * np.eye(3)), excitations=[4])
        wf, pgrad = generate.make_sj_pgrad(slater, pyqmc.JastrowSpin, rcut=7.5)
        generate.load_wf_from_linemin("../linemin_conv_spin.hdf", wf, n=20, nprim=1)

        ncore = 20
        nconfig = ncore * 30
        configs = pyqmc.initial_guess(slater.supercell, nconfig, r=1.0)

        cluster = LocalCluster(n_workers=ncore, threads_per_worker=1)
        client = Client(cluster)
    
        # Warm up VMC
        df, configs = vmc(wf, configs, nsteps=3, verbose=True, client=client)
        print("Finished warmup VMC", flush=True)

        # Optimize jastrow
        hdf_file="exc4_linemin.hdf"
        lm_dict = dict(hdf_file=hdf_file, verbose=True, client=client, )
        lm_dict.update(dict(vmcoptions={"nsteps":40}), maxiters=50)
        wf, lm_df = line_minimization(wf, configs, pgrad, **lm_dict)

I get the following warnings and eventual error

    /projects/wagner/anaconda3/envs/pyscf/lib/python3.7/contextlib.py:119: UserWarning: Creating scratch directories is taking a surprisingly long time. This is often due to running workers on a network file system. Consider specifying a local-directory to point workers to write scratch data to a local disk.
      next(self.gen)
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 2.48 GB -- Worker memory limit: 3.36 GB
    /projects/wagner/anaconda3/envs/pyscf/lib/python3.7/contextlib.py:119: UserWarning: Creating scratch directories is taking a surprisingly long time. This is often due to running workers on a network file system. Consider specifying a local-directory to point workers to write scratch data to a local disk.
      next(self.gen)
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 2.57 GB -- Worker memory limit: 3.36 GB
    ... (many more lines like this) ...
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 3.11 GB -- Worker memory limit: 3.36 GB
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 3.29 GB -- Worker memory limit: 3.36 GB
    distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
    distributed.nanny - WARNING - Restarting worker
    distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
    distributed.nanny - WARNING - Restarting worker
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 2.41 GB -- Worker memory limit: 3.36 GB
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 2.65 GB -- Worker memory limit: 3.36 GB
    distributed.worker - WARNING - Worker is at 85% memory usage. Pausing worker.  Process memory: 2.89 GB -- Worker memory limit: 3.36 GB
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 2.89 GB -- Worker memory limit: 3.36 GB
    distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 3.13 GB -- Worker memory limit: 3.36 GB
    distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
    Traceback (most recent call last):
      File "exc4_opt.py", line 29, in <module>
        wf, lm_df = line_minimization(wf, configs, pgrad, **lm_dict)
      File "/projects/wagner/wawheel2/code/pyqmc/pyqmc/dasktools.py", line 171, in line_minimization
        return pyqmc.line_minimization(*args, vmc=distvmc, lm=dist_lm_sampler, **kwargs)
      File "/projects/wagner/wawheel2/code/pyqmc/pyqmc/linemin.py", line 198, in line_minimization
        coords, last_en, pgrad, Sij, en, en_err = gradient_energy_function(x0, coords)
      File "/projects/wagner/wawheel2/code/pyqmc/pyqmc/linemin.py", line 177, in gradient_energy_function
        data, coords = vmc(wf, coords, accumulators={"pgrad": pgrad_acc}, **vmcoptions)
      File "/projects/wagner/wawheel2/code/pyqmc/pyqmc/dasktools.py", line 96, in distvmc
        allresults = list(zip(*[r.result() for r in runs]))
      File "/projects/wagner/wawheel2/code/pyqmc/pyqmc/dasktools.py", line 96, in <listcomp>
        allresults = list(zip(*[r.result() for r in runs]))
      File "/projects/wagner/anaconda3/envs/pyscf/lib/python3.7/site-packages/distributed/client.py", line 221, in result
        six.reraise(*result)
      File "/projects/wagner/anaconda3/envs/pyscf/lib/python3.7/site-packages/six.py", line 703, in reraise
        raise value
    distributed.scheduler.KilledWorker: ('vmc-8198f95b6e97c8f539177b02d94100a9', <Worker 'tcp://127.0.0.1:35107', memory: 0, processing: 15>)

It's clear that the problem is in running distvmc, and it seems to be in the PGradAccumulator. For now, I'm just documenting the problem.

**Comment by willwheelera** [comment]
Looks like I'm constructing a big array in Ewald.ewald_electron

        ei_distances = configs.dist.dist_i(self.atom_coords, configs.configs)
        rvec = ei_distances[:, :, :, np.newaxis, :] + self.lattice_displacements
        r = np.linalg.norm(rvec, axis=-1)
        ei_real_separated = np.einsum(
            "k,ijkl->ji", -self.atom_charges, erfc(self.alpha * r) / r 
        )

rvec has shape (256, 30, 64, 125, 3), which is just too big. And the same size will be reached for ee_distances. I seem to have fixed the problem by looping over the 125 lattice displacements, although it might be slower.

**Comment by lkwagner** [comment]
Can you try to remove the temporary variable rvec? For ee_distances, it should be significantly smaller, since we just return the minimal image by convention, right?

**Comment by willwheelera** [comment]
What do you mean by remove? From looking at print statements, it looks like the program never got past constructing rvec for ei. There are 256 electrons, so there are 255*128 electron pairs, which is double the 256*64 ei pairs. The lattice_displacements are for evaluating erfc for the real-space Ewald contribution, so it's not enough to just have the minimal image. nlatvec=2 by default, making 125 lattice displacements to sum over. Maybe for such a big cell nlatvec=1 would be enough? It doesn't seem like a permanent solution because I don't think it would work in a smaller cell geometry that still has a lot of electrons.

**Comment by lkwagner** [comment]
say you do 
```
a = np.einsum(i,j->ij, bigvec1, bigvec2
b = np.einsum(ij,j -> i,a,bigvec3)
```
Then `a` will be constructed and take up N^2 memory, then be multiplied by b. If you instead do 
```
b = np.einsum(i,j,j->i, bigvec1,bigvec2,bigvec3)
```
then you never construct a N^2 memory object. 

in the case above, you don't need to actually construct rvec or r. You could make it a single numpy call which would then get run in C.

**Comment by willwheelera** [comment]
With the norm and the erfc, it's not clear to me how I can use your suggestion.

I'm not sure how much impact the for loop has. I can try profiling it. I was also thinking if it's a big deal we can consider an optional cython function.

**Comment by lkwagner** [comment]
I agree that it's not as straightforward as my example. I think it's worth looking into solutions of this sort; it'd be nice to avoid having C code at all. I think numpy has some functionality of this sort at worst through the ufuncs but I don't have experience with them.

I also agree ( :) ) that a good stopgap is to insert a loop. For large numbers of walkers or electrons it shouldn't be a big deal.

**Comment by willwheelera** [comment]
The Ewald object knows how many electrons and atoms it has. We could set it to use the loop function if there is a high number of electrons or atoms, and otherwise use the vectorized one.

**Comment by lkwagner** [comment]
Sure, although I wouldn't do that until we see some performance reasons. The whole code doesn't perform well unless there are a lot of electrons or a lot of walkers.

**Comment by lkwagner** [comment]
Has this been resolved?

**Comment by willwheelera** [comment]
Yeah, this was fixed in PR #158 and I guess we decided to stick with the for loop for now
